src/main/java/mat/server/service/impl/FhirCqlLibraryServiceImpl.java

```diff
@@ -0,0 +1,158 @@
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.exolab.castor.mapping.MappingException;
+import org.exolab.castor.xml.MarshalException;
+import org.exolab.castor.xml.ValidationException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.io.IOException;
+import java.util.Optional;
+
+@Service
+public class FhirCqlLibraryServiceImpl implements FhirCqlLibraryService{
+
+    private static final Log logger = LogFactory.getLog(FhirCqlLibraryServiceImpl.class);
+
+    @Autowired
+    private FhirLibraryConversionRemoteCall fhirLibraryConversionRemoteCall;
+
+    @Autowired
+    private FhirConvertResultResponse fhirConvertResultResponse;
+
+    @Autowired
+    private CQLLibraryServiceInterface cqlLibraryServiceInterface;
+
+    @Autowired
+    private CQLLibraryDAO cqlLibraryDAO;
+
+    private final CqlParser cqlParser;
+
+    private final CqlVisitorFactory cqlVisitorFactory;
+
+    private final XMLMarshalUtil xmlMarshalUtil = new XMLMarshalUtil();
+
+    public FhirCqlLibraryServiceImpl(CqlParser cqlParser, CqlVisitorFactory cqlVisitorFactory) {
+        this.cqlParser = cqlParser;
+        this.cqlVisitorFactory = cqlVisitorFactory;
+    }
+
+    @Override
+    public FhirConvertResultResponse convertCqlLibrary(CQLLibraryDataSetObject sourceLibrary, String loggedInUser) throws MatException, MarshalException, MappingException, IOException, ValidationException {
+
+        if (!sourceLibrary.isFhirConvertible()) {
+            throw new MatException("Cql Library cannot be converted to FHIR");
+        }
+
+        fhirConvertResultResponse.setSourceLibrary(sourceLibrary);
+        cqlLibraryServiceInterface.isLibraryAvailableAndLogRecentActivity(sourceLibrary.getId(), loggedInUser);
+
+        CQLLibrary existingLibrary = cqlLibraryDAO.find(sourceLibrary.getId());
+
+        /*if draft FHIR library already exists then delete it */
+        deleteDraftFhirLibrariesInSet(existingLibrary.getSetId());
+
+        ConversionResultDto conversionResult = fhirLibraryConversionRemoteCall.convert(sourceLibrary.getId(), ConversionType.CONVERSION);
+        Optional<String> fhirCqlOpt = getFhirCql(conversionResult);
+
+        FhirValidationStatus validationStatus = createValidationStatus(conversionResult);
+        fhirConvertResultResponse.setValidationStatus(validationStatus);
+
+        if (!fhirCqlOpt.isPresent()) {
+            // If there is no FHIR CQL, then we don't persist the Library. FHIR Library cannot be created.
+            throw new MatException("Your library cannot be converted to FHIR. Outcome: " + validationStatus.getOutcome() + " Error Reason: " + validationStatus.getErrorReason());
+        } else {
+            /*Create a new cql Library object and store it in DB*/
+            persistFhirLibrary(loggedInUser, existingLibrary, fhirCqlOpt);
```

Comment:
at this point, it's known that fhirCqlOpt is present. you can pass unwrapped value with fhirCqlOpt.get